File: Filter.py
import math

class Filter:
    def __init__(self):
        self.prevVal = 0.0

    def filter(self, value, time):
        self.prevVal = value
        return value

class SimpleFilter(Filter):
    def __init__(self, innerRadius):
        super().__init__()
        self.iRadius = innerRadius

    def filter(self, value, time):
        if math.fabs(value - self.prevVal) <= self.iRadius:
            value = self.prevVal
        return super().filter(value, time)

class RadialFilter(SimpleFilter):
    def __init__(self, innerRadius, outerRadius, descentSpeed, smoothness):
        super().__init__(innerRadius)
        self.oRadius = outerRadius
        self.descentSpeed = descentSpeed
        self.smoothness = smoothness

    def filter(self, value, time):
        distance = self.prevVal - value

        #Cleaned up the math
        if math.fabs(distance) > self.iRadius and math.fabs(distance) <= self.oRadius:
            value = self.prevVal + (1 - self.descentSpeed) * self.oRadius * time * (math.atan(distance * smoothness) * 2/math.pi)
        
        return super().filter(value, time)

File: Filter.cpp
#include <cmath>

class Filter
{
protected:
    float prevVal;

public:
    virtual float filter(float value, float time) {
        prevVal = value;
        return value;
    }

    Filter() : prevVal(0) {}

    virtual ~Filter() {}
};

class SimpleFilter : public Filter
{
protected:
    float iRadius;

public:
    float filter(float value, float time) override {
        if (std::fabs(value - prevVal) < iRadius) {
            value = prevVal;
        }

        return Filter::filter(value, time);
    }

    SimpleFilter(float innerRadius) : Filter(), iRadius(innerRadius) {}
};


class RadialFilter : public SimpleFilter
{
protected:
    float oRadius;
    float descentSpeed;

public:
    float filter(float value, float time) override {
        float distance = prevVal - value;

        if (std::fabs(distance) < iRadius) {
            value = prevVal;
        }

        else if (std::fabs(distance) < oRadius) {  
            // Apply a curve that slows down as it gets closer
            //value = prevVal + distance * descentSpeed * time;  

            //Apply linear amount of movement (Should be a better way that doesnt need if statement)
            if (distance < 0)
            {
                value = prevVal + (1 - descentSpeed) * oRadius * time;
            }
            else
            {
                value = prevVal - (1 - descentSpeed) * oRadius * time;
            }
        }

        return Filter::filter(value, time);
    }

    RadialFilter(float innerRadius, float outerRadius, float descentSpeed) 
        : SimpleFilter(innerRadius), oRadius(outerRadius), descentSpeed(descentSpeed) {}
};
